marimo/_tutorials/fileformat.py

```diff
@@ -221,38 +206,352 @@ def echo(text):
+        carry this information in their signature. You can add the setup cell
+        from the general menu of the editor under: 💠 Add setup cell
+        """
+    )
+    return
+
+
+@app.cell(hide_code=True)
+def _():
+    mo.md(
+        r"""
+        ### Body Cell
+
+        This is the general, and most common type of cell type. A regular body
+        cell (like this one), is saved as follows:
+
+        ```python
+        @app.cell(hide_code=True) # <- Cell options are also saved to file
+        def _(increment, a): # <- Signatures are automatically generated and consist of used references
+            message = mo.md(f'''
+                `a` incremented is {increment(a)}
+            ''')
+            return (message,) # <- All used cell definitions are returned
+
+        ```
+
+        The default name of a cell is `_`, but you can explicitly give cells a
+        name, as seen in the following cell example below:
+        """
+    )
+    return
 
 
 @app.cell
-def __():
-    import marimo as mo
-    return mo,
+def cell_example():
+    runtime_definition = "Defined in a normal cell, executed in runtime."
+
+    print("This cell is a normal body cell!")
+
+    "Notice how the last line is output"
+    return (runtime_definition,)
+
+
+@app.cell(hide_code=True)
+def _():
+    mo.md(
+        r"""
+        ### Functions / Class cells
+
+        marimo directly exposes functions and classes that can be directly
+        serialized, opposed to wrapping definitions in a superfluous `@app.cell`.
+
+        In marimo versions older than `0.13.0`, cells with only a function definition looked like: 
+
+        ```python
+        @app.cell
+        def _():
+            def say_hello(name="World"):
+                return f"Hello, {name}"
+            return say_hello
+        ```
+        this is now rendered as:
+
+        ```python
+        @app.function
+        def say_hello(name="World"):
+            return f"Hello, {name}"
+        ```
+
+        See the following example in the playground below:
+        """
+    )
+    return
+
+
+@app.function
+def function_example():
+    """Notice how an indicator in the lower right corner shows this cell is serialized differently"""
+    return mo.ui.slider(1, 100)
+
+
+@app.cell(hide_code=True)
+def _():
+    mo.md(
+        r"""
+        likewise, standalone classes are also exposed with the
+        `@app.cell_definition` decorator:
+
+        ```python
+        @app.class_definition
+        class MyClass: ...
+        ```
+
+        Moreover, classes and functions can refer to each other like any other
+        python module. As a restriction of marimo, we don't allow cycles between
+        definitions- but recursion and directed references are fine. Consider
+        the following:
+        """
+    )
+    return
+
+
+@app.class_definition
+@dataclasses.dataclass
+class SettingExample:
+    """
+    TODO: I realized this is probably a poor example because
```

Comment:
Bookmarking this before merge, any suggestions? Worth "fixing" ui registry for?

Reply:
Thanks for the cleanup!

I really like the idea of a `@mo.ui.dataclass`, I don't think it'd be too hard to extend, and could enforce it on the global scale- but we would need public UI types.

I really wanted this functionality to share plots or types of plots across notebooks- which is still a reasonable example. I actually have a good one for log histograms that I'll put in